### PR TITLE
Sort output config fields stably

### LINKED ISSUE #4745: Output config options ordered randomly
## Expected Behavior

My plugin's output's config options are show in the same order as I added them in the output config class independent of their amount.

## Current Behavior

My plugin's output's config options are show in a random order if there are more than 10.

## Possible Solution

Use a stable sort algorithm while sorting by optionality.

## Steps to Reproduce (for bugs)

1. Create an output plugin with more than 10 non-optional config options.
2. View the config in the web UI.

## Context

@N-o-X and me are developing a Graylog plugin.

## Your Environment

* Graylog Version: 2.4.3
* Operating System: CentOS 7

### PR DESCRIPTION
**Pull request by Al2Klimov**
## Description

fixes #4745

## Motivation and Context

#4745

## How Has This Been Tested?

Not at all. Unfortunately I couldn't build Graylog independent of my change.

## Screenshots (if appropriate):

N/A

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by Al2Klimov** [comment]
PS: I (and @N-o-X and @bobapple) really would like you to include this in v2.4.x.

**Comment by Al2Klimov** [comment]
Unfortunately I seem not to have access to your Jenkins so I can't even figure out my mistake.

**Comment by Al2Klimov** [comment]
As discussed w/ @jalogisch (offline):

* Command executed by Jenkins (effectively): `node node_modules/.bin/eslint -c packages/graylog-web-plugin/.eslintrc -f checkstyle src/components/configurationforms/ConfigurationForm.jsx`
* I fixed my code style errors, but there are some more (of others)
* I'll open a dedicated PR for 2.4.x

File: graylog2-web-interface/src/components/configurationforms/ConfigurationForm.jsx [hunk 1]
```diff
@@ -153,9 +153,12 @@ const ConfigurationForm = createReactClass({
       shouldAutoFocus = false;
     }
 
-    const configFieldKeys = $.map(this.state.configFields, (v, k) => k).sort(this._sortByOptionality);
+    const configFieldKeys = $.map(this.state.configFields, (v, k) => k)
+      .map((v, i) => ({ key: v, pos: i }))
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/configurationforms/ConfigurationForm.jsx]: I know you didn't wrote that line, but I would vote for replacing the `$.map(this.state.configFields, (v, k) => k)` with the much clearer, in my opinion, `Object.keys(this.state.configFields)`.

Please also use a clearer name for `v`. I suggest `key` instead. In the end we minify the code, so better make it easier to understand.

File: graylog2-web-interface/src/components/configurationforms/ConfigurationForm.jsx [hunk 2]
```diff
@@ -83,7 +83,7 @@ const ConfigurationForm = createReactClass({
   },
 
   _sortByOptionality(x1, x2) {
-    return (this.state.configFields[x1].is_optional - this.state.configFields[x2].is_optional);
+    return this.state.configFields[x1.key].is_optional - this.state.configFields[x2.key].is_optional || x1.pos - x2.pos;
```
Comment (edmundoa) [inline graylog2-web-interface/src/components/configurationforms/ConfigurationForm.jsx]: I think we should rewrite this in a more explicit way.
Subtracting two booleans is already hard to understand without deep understanding of JS internals, and using the `or` operator to do another comparison when the previous result was 0 only makes it harder to understand.

**Comment by edmundoa** [comment]
I got carried on by the review and I forgot to mention: thank you very much for your contribution! Let's work through the code and get this fix for good 🙂

**Comment by Al2Klimov** [comment]
Hello @edmundoa and thank you for responding!

> we should use another strategy, for instance sorting (...) by key name.

This would be a breaking change for our output plugin and for all the others.
So far plugin authors registered the config options in the desired order (e.g. host, user, password, ...), but this would (1) shuffle all options (of all plugins) (2) during a patchlevel release.

Besides that.. AFAIK no one complained about that.

So I won't change that – at least not in this PR as it doesn't solve the problem.

But you can expect the other changes soon.

Best,
AK

**Comment by edmundoa** [comment]
@Al2Klimov sounds reasonable to me. For what I read, most browsers respect the `Object` property order, even if the standard does not require that. The way I see it, at least the situation will be better than it was before for most people.

To further guarantee that the inputs will appear in the order specified by the author (even if it is implicitly), we would need to explicitly export the order of each key in the API, allowing to read that property from the JS code and sorting by it. That change would be fine for 3.0, but it would be a no-go for 2.4.x.

**Comment by edmundoa** [comment]
@Al2Klimov to be honest, I don't know the reason for it and it probably happened organically. I _guess_ the reason was to let users see the mandatory fields they had to fill out first, in case they didn't pay much attention to the form. Also sometimes configuration fields are put together from different classes and maybe the sorting looked somehow "wrong".

Feel free to open an issue for that so we can discuss it and maybe also change it.

**Comment by edmundoa** [comment]
Thank you for your contribution @Al2Klimov! Do you want to update the 2.4 branch with the changes or should I?

**Comment by Al2Klimov** [comment]
I'll do.

**Pull request merged**